### ISSUE TITLE
Reunification of the DAQ run interface(s)

### ISSUE THREAD

**Issue by guiguem** [project8/dragonfly#69]
Currently every instrument has its own class "InstrumentAcquisitionInterface" which contains hardcoded endpoints to interact with.
This class inheritates from DAQ provider class.
Then the config file used to create the service of this DAQInterface uses the "InstrumentAcquisitionInterface" class as module.

Instead, we should directly use the DAQ provider class inside the config file: use more the provider endpoint of the instrument and less hardcoded endpoints.
This will always force us to define a common way to start run which would then be instrument independent as pointed out here: https://github.com/project8/hardware/wiki/DAQ_Components

**Comment by laroque** [comment]
I certainly agree that hard-coding endpoint names is very bad, they should be configurable.

The problem is that the interfaces to the DAQ systems are very much not uniform (for example taking RSA data for 60 seconds requires a start command, followed by a stop command 60 seconds later; in mantis, 60 seconds of data requires requesting 10 seconds of data 6 times in a row). There has to be a place where we apply complex logic, to get from the uniform instruction "take 60 seconds of data" to whatever sequence of configurations and commands are required to make that happen.

To date (and described on the wiki page linked above), the line that has been drawn is that the DAQ service is responsible for implementing core functionality in a way that is natural for a given DAQ system, and is not uniform. The DAQInterface classes then serve a role very similar to a MultiSet, of providing an easy/uniform interface and allowing the required logic to be stored programatically so that is it is always executed consistently. The methods in the `DAQProvider` class are the common way to start a run, their implementations are necessarily different because the systems are different.

If we want to move that logic out of the specific <DAQ>Interface classes then we would still end up needing to have another intermediate service for the RSA (taking a run involves using multiple commands, which means talking to multiple endpoints, which can't happen between endpoints within a service... see extended discussion in #57). For psyllid and mantis, if we ever want to use it again, we would have to either add a similar intermediate service or re-write the c++ code to implement a meta interface on top of the run-queue system. @nsoblath could better comment on how easy that might be.

**Comment by guiguem** [comment]
I agree with you that the way instruments are configured and start to save data is different.
This is why there is now (RSA and ROACH) a dragonfly layer which interacts with the instrument or any more basic class.
This layer is here to uniformize the interfacing with any other class.

The complex logic should be in the provider of the associated instrument: for the RSA, most of the logic currently lives there btw, except that more delocalization of the specificities of the instrument can still be done (such as the checks of the clock source).
The generic DAQProvider would then call several generic methods (start, end, check/configure if some more specific actions/check are required).
As far as I see, there already has to be a instrument provider (the ROACH has roach2interface.py which already contains a is_running method).
If it is not too complex to implement, it would be nice to only have a generic DAQProvider (which does all the tricks of metadata and sql grabbing) which is used by the config file.

**Comment by laroque** [comment]
Complex logic cannot be in the provider for any instrument. This is true for DAQ for the same reason that it is true for the ESR. It is not possible for one object in a service to send a dragonfly request to another object in the same service (this is the issue #57 discussion). Until someone has time to update the pika depenency to >=0.10.0 and rework the dripline backend to support asynchronous processing of requests via the asyncio package, this will not be possible and logic will have to be done in a dedicated service which is independent of the hardware service.

The presence of is_running in the roach2interface.py does not do what you think it does. This tells us if the roach itself is digitizing data, but that does not mean that we are taking data, it simply tells us that packets are available to psyllid. To collect data psyllid must also be running and must be in the midst of a run. These are independent components which must be coordinated.

The generic DAQProvider is already used, via class inheritance, in any other specific-DAQ interface. It does all of the "tricks of metadata and sql grabbing" as its implementation of the standard interface methods (start, end, check). It defines the universal components of what happens under each of those methods. The DAQ-specific subclasses then extend those methods to include the behaviors required for themselves. This makes particularly good sense because there are aspects of the metadata, which the DAQProvider needs access to in order to generate the context files properly, which are entirely dependent on the details of the particular DAQ system. We could split the SpecificDAQInterface services into two services, a generic RunMeta service and a DAQCoordination, but I don't see what we gain by doing so.
